### PR TITLE
[babel__traverse] Lots of fixes for babel traverse

### PR DESCRIPTION
**Pull request by danez**
See inline comments for sources.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/babel__traverse/index.d.ts [hunk 1]
```diff
@@ -106,7 +106,7 @@ export class Scope {
     /** Possibly generate a memoised identifier if it is not static and has consequences. */
     maybeGenerateMemoised(node: Node, dontPush?: boolean): t.Identifier;
 
-    checkBlockScopedCollisions(local: Node, kind: string, name: string, id: object): void;
+    checkBlockScopedCollisions(local: Binding, kind: BindingKind, name: string, id: object): void;
```
Comment (danez) [inline types/babel__traverse/index.d.ts:109]: https://github.com/babel/babel/blob/main/packages/babel-traverse/src/scope/index.ts#L583-L588

File: types/babel__traverse/index.d.ts [hunk 2]
```diff
@@ -180,23 +180,30 @@ export class Scope {
     removeBinding(name: string): void;
 }
 
+export type BindingKind = 'var' | 'let' | 'const' | 'module' | 'hoisted' | 'param' | 'local' | 'unknown';
+
```
Comment (danez) [inline types/babel__traverse/index.d.ts:184]: https://github.com/babel/babel/blob/main/packages/babel-traverse/src/scope/binding.ts#L5-L13

File: types/babel__traverse/index.d.ts [hunk 3]
```diff
@@ -180,23 +180,30 @@ export class Scope {
...
 export class Binding {
-    constructor(opts: {
-        existing: Binding;
-        identifier: t.Identifier;
-        scope: Scope;
-        path: NodePath;
-        kind: 'var' | 'let' | 'const';
-    });
+    constructor(opts: { identifier: t.Identifier; scope: Scope; path: NodePath; kind: BindingKind });
```
Comment (danez) [inline types/babel__traverse/index.d.ts:186]: https://github.com/babel/babel/blob/main/packages/babel-traverse/src/scope/binding.ts#L31-L41

File: types/babel__traverse/index.d.ts [hunk 4]
```diff
@@ -180,23 +180,30 @@ export class Scope {
...
     identifier: t.Identifier;
     scope: Scope;
     path: NodePath;
-    kind: 'var' | 'let' | 'const' | 'module';
+    kind: BindingKind;
```
Comment (danez) [inline types/babel__traverse/index.d.ts:190]: https://github.com/babel/babel/blob/main/packages/babel-traverse/src/scope/binding.ts#L29

File: types/babel__traverse/index.d.ts [hunk 5]
```diff
@@ -180,23 +180,30 @@ export class Scope {
...
+    hasValue?: boolean;
+    value?: any;
+
+    deopValue(): void;
+    setValue(value: any): void;
+    clearValue(): void;
+
+    reassign(path: NodePath): void;
+    reference(path: NodePath): void;
+    dereference(): void;
```
Comment (danez) [inline types/babel__traverse/index.d.ts:206]: https://github.com/babel/babel/blob/main/packages/babel-traverse/src/scope/binding.ts#L57-L110

File: types/babel__traverse/index.d.ts [hunk 6]
```diff
@@ -495,7 +502,7 @@ export class NodePath<T = Node> {
 
     setScope(): void;
 
-    setContext(context: TraversalContext): NodePath<T>;
+    setContext(context?: TraversalContext): this;
```
Comment (danez) [inline types/babel__traverse/index.d.ts:505]: https://github.com/babel/babel/blob/main/packages/babel-traverse/src/path/context.ts#L156-L159

File: types/babel__traverse/index.d.ts [hunk 7]
```diff
@@ -560,9 +567,17 @@ export class NodePath<T = Node> {
...
+
+    getOuterBindingIdentifiers(duplicates: true): Record<string, t.Identifier[]>;
+    getOuterBindingIdentifiers(duplicates?: false): Record<string, t.Identifier>;
+
+    getBindingIdentifierPaths(duplicates: true, outerOnly?: boolean): Record<string, Array<NodePath<t.Identifier>>>;
+    getBindingIdentifierPaths(duplicates?: false, outerOnly?: boolean): Record<string, NodePath<t.Identifier>>;
 
-    getOuterBindingIdentifiers(duplicates?: boolean): Node[];
+    getOuterBindingIdentifierPaths(duplicates: true): Record<string, Array<NodePath<t.Identifier>>>;
+    getOuterBindingIdentifierPaths(duplicates?: false): Record<string, NodePath<t.Identifier>>;
```
Comment (danez) [inline types/babel__traverse/index.d.ts:580]: https://github.com/babel/babel/blob/main/packages/babel-traverse/src/path/family.ts#L570-L575

File: types/babel__traverse/index.d.ts [hunk 8]
```diff
@@ -560,9 +567,17 @@ export class NodePath<T = Node> {
         : never;
     get(key: string, context?: boolean | TraversalContext): NodePath | NodePath[];
 
-    getBindingIdentifiers(duplicates?: boolean): Node[];
+    getBindingIdentifiers(duplicates: true): Record<string, t.Identifier[]>;
+    getBindingIdentifiers(duplicates?: false): Record<string, t.Identifier>;
```
Comment (danez) [inline types/babel__traverse/index.d.ts:571]: https://github.com/babel/babel/blob/main/packages/babel-traverse/src/path/family.ts#L463-L471

File: types/babel__traverse/index.d.ts [hunk 9]
```diff
@@ -560,9 +567,17 @@ export class NodePath<T = Node> {
...
+
+    getOuterBindingIdentifiers(duplicates: true): Record<string, t.Identifier[]>;
+    getOuterBindingIdentifiers(duplicates?: false): Record<string, t.Identifier>;
```
Comment (danez) [inline types/babel__traverse/index.d.ts:574]: https://github.com/babel/babel/blob/main/packages/babel-traverse/src/path/family.ts#L482-L490

File: types/babel__traverse/index.d.ts [hunk 10]
```diff
@@ -560,9 +567,17 @@ export class NodePath<T = Node> {
...
+
+    getBindingIdentifierPaths(duplicates: true, outerOnly?: boolean): Record<string, Array<NodePath<t.Identifier>>>;
+    getBindingIdentifierPaths(duplicates?: false, outerOnly?: boolean): Record<string, NodePath<t.Identifier>>;
```
Comment (danez) [inline types/babel__traverse/index.d.ts:577]: https://github.com/babel/babel/blob/main/packages/babel-traverse/src/path/family.ts#L504-L511

**Review by marvinhagemeister — APPROVED** [review]

**Review by marvinhagemeister — APPROVED** [review]

**Review by andrewbranch — APPROVED** [review]

**Comment by danez** [comment]
Ready to merge

**Pull request merged**